### PR TITLE
Updated to work on windows.

### PR DESCRIPTION
**Pull request by Vethanis**

### CODE REVIEW

**Comment by martymac** [comment]
Hello,

Thanks for your contribution.

This is a huge patch... and several lines have been reformatted too, making it quite hard to follow.

Could you split it into multiple commits with explicit commit logs ?

Also, if the aim is to make it work on Windows, can you be more specific about the parts that currently do not work on that OS by submitting dedicated commits/patches ?

Ganael.

**Comment by darealshinji** [comment]
To compile on Windows you only need to change a few includes. To compile on MSVC you also need to define the `S_ISDIR()` macro and to add an implementation of getopt ~~which you can find here (LGPL): https://github.com/libimobiledevice-win32/getopt~~ which you can take from mingw-w64:
https://sourceforge.net/p/mingw-w64/mingw-w64/ci/master/tree/mingw-w64-headers/crt/getopt.h
https://sourceforge.net/p/mingw-w64/mingw-w64/ci/master/tree/mingw-w64-crt/misc/getopt.c

diff:
``` diff
--- grpar.c
+++ grpar.c
@@ -55,8 +55,15 @@

 /* read(2), getopt(3) */
 #include <sys/types.h>
-#include <sys/uio.h>
-#include <unistd.h>
+#if defined(_WIN32)
+#  include <getopt.h>
+#  ifndef S_ISDIR
+#    define S_ISDIR(mode)  (((mode) & S_IFMT) == S_IFDIR)
+#  endif
+#else
+#  include <sys/uio.h>
+#  include <unistd.h>
+#endif

 #if defined(__FreeBSD__)
   /* le32toh(9) */
```

Compiling with Mingw: `i686-w64-mingw32-gcc -Wall -O2 grpar.c -o grpar.exe -s`
Compiling with MSVC: `cl /O2 /I. grpar.c getopt.c`

**Comment by martymac** [comment]
Hello,

Thanks for the followup. Would that (much simpler) patch be enough to build and make grpar useable on Windows ?

**Comment by darealshinji** [comment]
It builds but now I get these errors:
>group archive header truncated
>error reading group archive TOC

I'll see if I can figure it out.

Update:
The issue was the missing `O_BINARY` flag.

**Comment by martymac** [comment]
Good :) Can you update the merge request with that patch (only) so that I can merge it ?

**Comment by martymac** [comment]
Re-submitted as PR#4
